drafts/multiformats.txt

```diff
@@ -0,0 +1,1736 @@
+
+4.1.  Examples
+
+   Given the sample text, "Multibase is awesome! \o/", the following
+   example Multibase encodings may be generated.
+
+   +------------+-----------------------------------------------------+
+   | Base       | Multibase                                           |
+   +------------+-----------------------------------------------------+
+   | 16 (upper) | F4D756C74696261736520697320617765736F6D6521205C6F2F |
+   | 16 (lower) | f4d756c74696261736520697320617765736f6d6521205c6f2f |
+   | 32         | BJV2WY5DJMJQXGZJANFZSAYLXMVZW63LFEEQFY3ZP           |
+   | 58         | zYAjKoNbau5KiqmHPmSxYCvn66dA1vLmwbt                 |
+   | 64         | MTXVsdGliYXNlIGlzIGF3ZXNvbWUhIFxvLw==               |
+   +------------+-----------------------------------------------------+
+
+4.2.  Multibase ABNF
+
+   ABNF
+
+   base-identifier =
+   multibase = base-identifier {base-encoded-data}
+
+
+
+
+
+
+
+
+Snell                    Expires January 2, 2020                [Page 7]
+
+Internet-Draft                Multiformats                     July 2019
+
+
+5.  Security Considerations
+
+   There are no additional security considerations introduced by the use
+   of Multiformat data formats.  Note, however, that various hash
+   functions that may be used with multiformats may be vulnerable to
+   various known security issues and limitations (e.g. sha1) and thus
+   SHOULD be avoided.
+
+6.  IANA Considerations
+
+6.1.  The Codec Registry
+
+6.1.1.  Codec Registration
+
+   Codec identifiers can be registered using the procedure described
+   herein.
+
+   Codecs are registered using the Expert Review policy (see Section 4.5
+   of [RFC8126]).  The goal of the registry is to reflect common use of
+   codecs on the Internet.  Therefore, the expert(s) should be strongly
+   biased towards approving registrations, unless they are abusive,
+   frivolous, not likely to be used on the Internet, or actively harmful
+   to the Internet and/or the Web (not merely aesthetically displeasing
+   or architecturally dubious).  Expert(s) can withhold registration of
+   codecs that are too general for the proposal application.
+
+   Expert(s) will clearly identify any issues that cause a registration
+   to be refused.  Advice about the semantics of a proposed codec can be
+   given, but if it does not block registration, this should be
+   explicitly stated.
+
+   When a request if approved, the expert(s) shall inform IANA, and the
+   regstration will be processed.  The IESG is the final arbiter of any
+   objection.
+
+   The Codec Registry is located at https://www.iana.org/assignments/
+   multiformat-codecs.  Registration requests can be made by following
+   instructions located there or by sending an email to the multiformat-
+   codecs@ietf.org mailing list.
+
+   Registration requests consist of at least the following information:
+
+   o  *Codec Name*: The name of codec
+
+   o  *Type*: The type of applicable multiformat
```

Comment:
We're not setting limits in here on type, and the initial registry contains lots of types not addressed in this spec so this is probably going to get us into trouble with new types coming along that aren't within the anticipated scope--which is fine, but we're not setting a limiting principle on valid types. And how is the IETF going to react to a doc that outlines two "multiformat" types but contains a registry that expands to other types that have unexplained identifiers ("ipld, holochain, wut? multiaddr, why isn't this specified in here?")

Do we need to call out the currently acceptable "types" and then have a separate procedure to introduce additional types, even if it's roughly similar to the above. Section 1 puts some rough boundaries on what multiformat can be, but I'm sure there will be attempts to expand that in very novel directions. What is a "dataformat", what does "compact" and "packed" really mean especially since multihashes can be arbitrarily large, so how about ELF files? Or any other file that has a magic number at the start that kind of looks like what other multiformats do with varints.

Reply:
It’s my understanding that the IETF will give us an IANA registry for multiformat entries and that sets up the governance on how new types are added and maintained.

Reply:
Yep, we can get the registry, but @rvagg is right about the initial contents. We're saying that there should be a spec for everything added but there's a lot there already that's not accounted for yet. We'll likely need to expand on that a bit and will definitely need to explain the other pieces. One approach we can take is to separate out the non multiformat and multihash entries into a separate document

Reply:
/cc @vmx

Reply:
Yes, that's pretty much it exactly. What we *could* also do, however, is separate out the additional entries into a separate document that adds them from the start.

Reply:
I've started separating this out. Will have more done by tomorrow.